blob/service.go

```diff
@@ -351,123 +362,187 @@ func (s *Service) retrieve(
 	headerGetterSpan.AddEvent("received eds", trace.WithAttributes(
 		attribute.Int64("eds-size", int64(len(header.DAH.RowRoots)))))
 
-	rowIndex := -1
+	// find the index of the row where the blob could start
+	inclusiveNamespaceStartRowIndex := -1
 	for i, row := range header.DAH.RowRoots {
 		if !namespace.IsOutsideRange(row, row) {
-			rowIndex = i
+			inclusiveNamespaceStartRowIndex = i
+			break
+		}
+	}
+	if inclusiveNamespaceStartRowIndex == -1 {
+		return nil, nil, ErrBlobNotFound
+	}
+
+	// end exclusive index of the row root containing the namespace
+	exclusiveNamespaceEndRowIndex := inclusiveNamespaceStartRowIndex
+	for i, row := range header.DAH.RowRoots[inclusiveNamespaceStartRowIndex:] {
+		if namespace.IsOutsideRange(row, row) {
+			exclusiveNamespaceEndRowIndex = inclusiveNamespaceStartRowIndex + i
 			break
 		}
 	}
 
-	getCtx, getSharesSpan := tracer.Start(ctx, "get-shares-by-namespace")
+	// calculate the square size
+	squareSize := len(header.DAH.RowRoots) / 2
 
-	// collect shares for the requested namespace
-	namespacedShares, err := s.shareGetter.GetSharesByNamespace(getCtx, header, namespace)
-	if err != nil {
-		if errors.Is(err, share.ErrNotFound) {
-			err = ErrBlobNotFound
+	// get all the shares of the rows containing the namespace
+	_, getSharesSpan := tracer.Start(ctx, "get-all-shares-in-namespace")
+	// store the ODS shares of the rows containing the blob
+	rowsShares := make([]share.Share, 0, (exclusiveNamespaceEndRowIndex-inclusiveNamespaceStartRowIndex)*squareSize)
+	// store the EDS shares of the rows containing the blob
+	rowsWithParityShares := make([][]shares.Share, exclusiveNamespaceEndRowIndex-inclusiveNamespaceStartRowIndex)
+	for rowIndex := inclusiveNamespaceStartRowIndex; rowIndex < exclusiveNamespaceEndRowIndex; rowIndex++ {
+		for colIndex := 0; colIndex < squareSize*2; colIndex++ {
+			share, err := s.shareGetter.GetShare(ctx, header, rowIndex, colIndex)
```

Comment:
having a blob that consists of 256 shares(or in case of GetAll request), you will make 256 independent network calls in case of the light node. Can we rework it by replacing with GetEDS instead?

Reply:
We should then define a threshold that we're over it, we should just get the whole EDS. Otherwise, we will be querying the whole EDS for any single blob even if it's just 1 row...

Reply:
Fetching the whole EDS will take more bandwidth, but it will be substantially faster. If we want to keep assuming the LN cannot afford that even for a few months from now, then we can do it by shares, but it will be slower and lead to more load on FN/BN

Note that this is based on our understanding and intuition of how the current protocol works, and we don't have a proper benchmark to prove it.

Reply:
I assume there is no way to do it over `SharesByNamespace`, right?

Reply:
`SharesByNamespace` doesn't provide enough information to generate the proof.

We could set a limit to 100 shares, for example. If the queried shares are more, then we query the EDS directly, otherwise we query the shares separately. What do you think?

Later on, we can improve on the value based on the blobs sizes that are submitted to Celestia and how often their proofs are queried.

Reply:
I'd any advocate for `GetEDS`. Assuming you have an EDS with blobs[b1,b2,b3,b4]. So in the worst case, you'd have to download the whole EDS to get b4, in the best case - only n shares of the first blob. I agree that GetEDS will take a lot more bandwidth in case of the best case, but imo it will be much much faster than the worst case. Also, GetEDS fits more for the GetAll request, since you will anyway have to fetch al the shares under the same namespace. We should also think about users that are actively using `GetAll`: getting shares one by one might take some time, so there might cases when the request will be timeouted.
Setting a threshold for N shares IMO is not the case(and the such logic will be harder to understand and maintain, which could lead a ton of bugs. As Blob Module is pretty hot and popular within the node, it will be painful to fix those bugs), since you don't know whether the share belongs to the blob that you are looking for(you will detect it once download all shares of the blob a create a commitment).

I personally think that should sacrifice the bandwidth in favour of speed.

Reply:
sure, as you prefer. This means that we don't need to improve the `GetShareProof` to only get the needed shares :D I'll update the implemenetation

Reply:
[2d209ad](https://github.com/celestiaorg/celestia-node/pull/3610/commits/2d209adef1e5cd662efe2994b60236a95ca76d4e)